### PR TITLE
Modify lcov.sh to work in tf-psa-crypto as well

### PR DESCRIPTION
**Pull request by tom-daubney-arm**
## Description

This PR solves part one of https://github.com/Mbed-TLS/TF-PSA-Crypto/issues/41.

Add repository detection (credit to davidhorstmann-arm for adding this in all.sh previously) and use repository detection to set the library directory and title variables.

I have tested this script locally against both mbedtls and tf-psa-crypto and the results were as expected.

Limitations: The modifications rely on the build directory being named "build" in tf-psa-crypto which may not always be the case. Perhaps this is ok for now, but I will document it. Otherwise perhaps we could supply the build directory name as a command line argument. Open to suggestions.

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [x] **changelog** not required (non user-visible change)
- [x] **backport** not required, new feature.
- [x] **tests** existing testing is sufficient.

## Notes for the submitter

Please refer to the [contributing guidelines](https://github.com/Mbed-TLS/mbedtls/blob/development/CONTRIBUTING.md), especially the
checklist for PR contributors.

### CODE REVIEW

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]

File: scripts/lcov.sh [hunk 1]
```diff
@@ -42,33 +42,46 @@ EOF
 
 set -eu
 
+# Repository detection
+in_mbedtls_repo () {
```
[the remark this answers is not in the document]
Comment (tom-daubney-arm) [reply scripts/lcov.sh:46]: Actually, having reread this I am a little bit confused about exactly what is needed. My understanding of the `in_mbedtls_repo` function is that it looks to see if all the named directories are present in the current working directory, and this includes looking for the `library` directory. So does that not effectively do the same thing as checking if `library` is present in the directory that the script is called from? I am probably missing some subtlety here but not sure what.

File: scripts/lcov.sh [hunk 2]
```diff
@@ -42,33 +42,46 @@ EOF
...
 if [ $# -gt 0 ] && [ "$1" = "--help" ]; then
     help
     exit
 fi
 
+if in_mbedtls_repo; then
+    library_dir='library'
+    title='Mbed TLS'
+else
+    library_dir='build/core'
```
Comment (ronald-cron-arm) [inline scripts/lcov.sh:81]:
```suggestion
    library_dir='core'
```
the script should be called from the build directory.

File: scripts/lcov.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (ronald-cron-arm) [reply scripts/lcov.sh:46]: Actually it is probably just the name that is confusing eventually. Here we do not want to check that we are at the root of the Mbed TLS or TF-PSA-Crypto repository. We want to detect that we are in an Mbed TLS of TF-PSA-Crypto build directory. That's the root of the repository when we do an in-tree build but not in the case of an out-of-tree build.
Then currently out-of-tree builds can be done only with CMake and looking at the directories in the build directory after running cmake to build the make files the directories that we use to check for "root detection" are all there.

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]

File: scripts/lcov.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply scripts/lcov.sh:46]: The job of this function is to determine whether the object files are in a directory called `library` or in `core`. And incidentally to set the product name in the report. Looking at other directories is irrelevant.

**Comment by gilles-peskine-arm** [comment]
> The modifications rely on the build directory being named "build" in tf-psa-crypto which may not always be the case

The script is supposed to be invoked from the build directory. It doesn't need to know the name of the build directory. The name of the build directory can be arbitrary; it's common to have multiple build directories with different configurations or compiler options.

File: scripts/lcov.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (tom-daubney-arm) [reply scripts/lcov.sh:46]: I see, that makes more sense now. I was trying to actually determine if we're in `mbedtls` or `tf-psa-crypto` but now see that that is not really what I should be trying to do

**Comment by tom-daubney-arm** [comment]
> > The modifications rely on the build directory being named "build" in tf-psa-crypto which may not always be the case
> 
> The script is supposed to be invoked from the build directory. It doesn't need to know the name of the build directory. The name of the build directory can be arbitrary; it's common to have multiple build directories with different configurations or compiler options.

Roger, thank you. I was confused over that but makes sense now. Will modify it.

File: scripts/lcov.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply scripts/lcov.sh:46]: You do need to determine whether you're in mbedtls or tf-psa-crypto. But that's a build directory of one of those two projects, not a source directory.

[the remark this answers is not in the document]
Comment (tom-daubney-arm) [reply scripts/lcov.sh:46]: Yes, I agree, I wasn't precise enough with that last comment. I meant I was trying to detect the root dir, rather than build

**Comment by tom-daubney-arm** [comment]
I've just noticed that the `gendesc` step of `lcov.sh` is failing on the psa repository, but not on mbedtls. I need to figure out why and push a change before this can be reviewed.

**Comment by tom-daubney-arm** [comment]
> I've just noticed that the `gendesc` step of `lcov.sh` is failing on the psa repository, but not on mbedtls. I need to figure out why and push a change before this can be reviewed.

The `tests/Descriptions.txt` needed copying over from `mbedtls` into `tf-psa-crytpo`. Once this was done the coverage report was produced correctly, so no modifications are required to this script and it can now be reviewed.

Thanks to @davidhorstmann-arm for helping me to get to the bottom of that.

**Review by ronald-cron-arm — APPROVED** [review]
LGTM, thanks.

**Review by gabor-mezei-arm — APPROVED** [review]
LGTM

**Pull request merged**